jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/UrlResourceFactory.java

```diff
@@ -0,0 +1,240 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2022 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util.resource;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.Path;
+import java.time.Instant;
+
+import org.eclipse.jetty.util.FileID;
+
+/**
+ * {@link ResourceFactory} for {@link java.net.URL} based resources.
+ */
+class UrlResourceFactory implements ResourceFactory
+{
+    private final String supportedProtocol;
+    private int connectTimeout;
+    private boolean useCaches;
+
+    protected UrlResourceFactory(String protocol)
```

Comment:
@joakime said:
> I'm leaving the check there.

Sorry, but I still don't agree with having the protocol here.  

> The protocol will be used ....
> 
> * to ensure that an old `ResourceFactory` instance isn't being reused for a different protocol.

But that is controlled by the `Index` of scheme to factory.  We don't need to encode checks that index works.

> * for output on `toString()` to indicate the supported protocol.

That is only meaningful in context.  In context we'd be dumping the Index with it's full mapping of scheme to factory, so such a dump will now contain the scheme twice: once in the key and then again in the value.

> * for configuring the URLConnection via protocol based system properties (usecaches, connecttimeout, etc).

But is is not?

> 
> There's no harm in having multiple `UrlResourceFactory` implementations, one for each requested protocol, it's a support layer for old school behaviors anyway.

It is not "old school" yet.   It is currently the only supported mechanism for several schemes.  Besides, "old school" is not an excuse for not doing it right.

Having the protocol in this class prohibits it ever being used as a catch-all.  I'm not sure we need a catch-all, but it is conceivable. If we wanted one, then with this needless protocol check, we'd have to iterator over all the URL handlers registered and register a factory for each.  Or we could just trust that `Map` works and do away with a needless check.

Reply:
> > * for configuring the URLConnection via protocol based system properties (usecaches, connecttimeout, etc).
> 
> But is is not?

Don't follow you on this one?

> Having the protocol in this class prohibits it ever being used as a catch-all. I'm not sure we need a catch-all, but it is conceivable. If we wanted one, then with this needless protocol check, we'd have to iterator over all the URL handlers registered and register a factory for each. Or we could just trust that Map works and do away with a needless check.

A catch-all is dangerous, and should never be a solution we implement.
And we have an `Index` controlling use/access to these implementation already.

Reply:
> > > * for configuring the URLConnection via protocol based system properties (usecaches, connecttimeout, etc).
> > 
> > 
> > But is is not?
> 
> Don't follow you on this one?

You don't use it for configuring the "URLConnection via protocol bases system properties".  The System properties used are class name based.  Not sure we should be using any system properties for configuration anyway!